File: src/api/schemas/folders.py
"""Folder navigation schema (API Contract v1.1, section 12)."""

from __future__ import annotations

from pydantic import BaseModel


class FolderOut(BaseModel):
    """One folder in the shared folder's visible structure.

    `path` and `name` are separate on purpose. `path` is the identity -- the
    canonical, reversible relative path -- and is what a client sends back as
    `folder_path`. `name` is the last segment rendered for reading, which for a
    folder created outside UTF-8 differs from the path entirely.

    A client must never rebuild a path by joining names. Send back the `path`
    this response gave you.

    Not present, deliberately: any server absolute path. Every value here is
    relative to the shared root, which the client never learns the location of.
    """

    path: str
    name: str
    parent_path: str | None = None
    depth: int
    #: Documents anywhere in this folder's subtree that the caller may read.
    document_count: int


class FolderListResponse(BaseModel):
    """A flat list; the client assembles the tree from `parent_path`.

    Flat rather than nested so a large tree can be rendered incrementally and
    a single folder's row can be refreshed without re-fetching its subtree.

    No pagination: the tree is navigation, and a partial tree is not navigable.
    """

    items: list[FolderOut]

    #: Every document the caller may browse, whether or not it sits in a
    #: folder. Not the sum of the depth-1 `document_count` values: a document
    #: at the top of the shared folder belongs to no folder and appears in no
    #: row, so summing would under-report -- and a shared folder with no
    #: subdirectories at all would report zero while holding documents.
    #:
    #: Lets a client draw a root row for the shared folder itself. The root has
    #: no path: selecting it means "no folder filter", which is the absence of
    #: the `folder_path` parameter rather than some value for it.
    total_documents: int = 0

File: frontend/src/api/folders.ts
import { getJson, type RequestOptions } from './client'

export interface FolderNode {
  /**
   * Canonical relative path -- the folder's identity, and what goes back to
   * the server as `folder_path`.
   *
   * For a folder created outside UTF-8 this is escaped and unreadable; that is
   * expected. Never build one by joining `name`s: the result would match no
   * document. Keep the value the server gave you and send it back verbatim.
   */
  path: string
  /** The last segment, rendered for reading. Display only. */
  name: string
  parent_path: string | null
  depth: number
  /** Readable documents anywhere beneath this folder. */
  document_count: number
}

export interface FolderListResponse {
  items: FolderNode[]
  /**
   * Every document the caller may browse, in a folder or not.
   *
   * Not the sum of the top-level `document_count` values: a document sitting
   * at the top of the shared folder belongs to no folder and appears in no
   * item, so summing would under-report -- and a shared folder with no
   * subdirectories would report zero while holding documents.
   */
  total_documents: number
  /**
   * Documents belonging to no folder -- the ones at the top of the shared
   * folder. They appear in no `items` entry, so without this the sidebar can
   * show the folders and the total but not account for the difference.
   *
   * Selected with `top_level_only`, not a `folder_path`: every path starts at
   * the root, so no prefix picks out exactly the documents not under one.
   */
  top_level_documents: number
}

/**
 * The shared folder's structure, as far as the caller may see it.
 *
 * Takes no filters on purpose: the tree is navigation, so it stays stable
 * while year, document kind and file type change. A folder holding only
 * manuals does not disappear when the user filters for reports -- selecting it
 * simply returns nothing.
 */
export function fetchFolders(options: RequestOptions = {}): Promise<FolderListResponse> {
  return getJson<FolderListResponse>('/folders', options)
}

export interface FolderTreeNode extends FolderNode {
  children: FolderTreeNode[]
}

/**
 * Assemble the flat list into a tree using `parent_path`.
 *
 * A node whose parent is missing is attached at the root rather than dropped:
 * the parent can only be absent if it held nothing readable, and hiding the
 * child as well would lose a folder the user is allowed to see.
 */
export function buildTree(items: FolderNode[]): FolderTreeNode[] {
  const byPath = new Map<string, FolderTreeNode>()
  for (const item of items) byPath.set(item.path, { ...item, children: [] })

  const roots: FolderTreeNode[] = []
  for (const node of byPath.values()) {
    const parent = node.parent_path == null ? undefined : byPath.get(node.parent_path)
    if (parent) parent.children.push(node)
    else roots.push(node)
  }

  const sortNodes = (nodes: FolderTreeNode[]): void => {
    nodes.sort((a, b) => a.name.localeCompare(b.name, 'ko'))
    for (const node of nodes) sortNodes(node.children)
  }
  sortNodes(roots)
  return roots
}

/** Every ancestor path of `path`, nearest last -- used to auto-expand a selection. */
export function ancestorPaths(items: FolderNode[], path: string): string[] {
  const byPath = new Map(items.map((item) => [item.path, item]))
  const chain: string[] = []
  let current = byPath.get(path)?.parent_path ?? null
  while (current) {
    chain.unshift(current)
    current = byPath.get(current)?.parent_path ?? null
  }
  return chain
}
